scripts/staging/fm-binclass.dml

```diff
@@ -0,0 +1,173 @@
+source("staging/fm.dml") as fm
+source("nn/layers/log_loss.dml") as log_loss
+source("nn/layers/sigmoid.dml") as sigmoid
+source("nn/layers/l2_reg.dml") as l2_reg
+source("nn/layers/cross_entropy_loss.dml") as cross_entropy_loss
+
+train = function(matrix[double] X, matrix[double] y, matrix[double] X_val, matrix[double] y_val)
+    return (matrix[double] w0, matrix[double] W, matrix[double] V, matrix[double] loss) {
+  /*
+   * Trains the FM model.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - y : label corresponds to n examples  
+   *  - lambda : regularization (5e-04)
+   * 
+   * Outputs:
+   *  - w0, W, V : updated model parameters.
+   *  - loss : computed loss with log_loss.
+   */
+
+  /* input propagation through layers
+   * fm::init -> adam::init -> fm::forward -> sigmoid::forward -> log_loss::forward \
+   *          adam::update <- fm::backward <- sigmoid::backward <- log_loss::backward <-
+   */
+
+ 
+    # 1.initialize fm core
+    [w0, W, V] = fm::init(d, k);
+    
+    # 2.initialize adam optimizer
+    ## Default values for some parameters
+    lr      = 0.001; 
+    beta1   = 0.9;       # [0, 1)
+    beta2   = 0.999;     # [0, 1)
+    epsilon = 0.00000001; 
+    t       = 0;
+    
+    # [mX, vX] = adam::init(X); # incase, you want to optimize input.
+    [mw0, vw0] = adam::init(w0);
+    [mW, vW]   = adam::init(W);
+    [mV, vV]   = adam::init(V);
+    
+    # Regularization
+    lambda = 5e-04
+    
+    # Optimize
+    print("Starting optimization")
+    batch_size = 64
+    iters = ceil(N / batch_size)
+    for (e in 1:epochs) {
+      for (i in 1:iters) {
+        # Get the next batch
+        beg = ((i-1) * batch_size) %% N + 1
+        end = min(N, beg + batch_size - 1)
+        X_batch = X[beg:end,]
+        y_batch = y[beg:end,]
+    
+        # 3.Send inputs through fm::forward
+        y = fm::forward(X_batch, w0, W, V);  
+    
+        # 4.Send the above result through sigmoid::forward
+        sfy = sigmoid::forward(y);
+    
+        # 5.Send the above result through log_loss::forward
+        loss = log_loss::forward(sfy);
+        
+        # Compute loss & accuracy for training & validation data every 100 iterations.
+        if (i %% 100 == 0) {
+          # Compute training loss & accuracy
+          loss_data   = cross_entropy_loss::forward(loss, y_batch);
+          loss_reg_X  = l2_reg::forward(X, lambda); 
+          loss_reg_w0 = l2_reg::forward(w0, lambda);
+          loss_reg_W  = l2_reg::forward(W, lambda);
+          loss_reg_V  = l2_reg::forward(V, lambda);
+          
+          accuracy = (rowIndexMax(loss) == rowIndexMax(y_batch));
+          loss = loss_data + loss_reg_X + loss_reg_w0 + loss_reg_W + loss_reg_V;
+          
+          # Compute validation loss & accuracy
+          probs_val = predict(X_val)
+          loss_val = cross_entropy_loss::forward(probs_val, y_val)
+          accuracy_val = mean(rowIndexMax(probs) == rowIndexMax(y_val))
```

Comment:
In a multi-class problem with one-hot encoding, this will work.  However, for a binary classification problem, you'll need to instead convert the `probs` to discrete values of either `0`, or `1` by either rounding or setting an arbitrary threshold.  Then, compare if those values are equal to the correct values in `y_val`, which will already be values of either `0` or `1`.  No `rowIndexMax` needed.